load() and DOM APIs

zslayton opened the pull request:
*Issue #, if available:* #531 

*Description of changes:*

Currently, `ion-js` only provides a streaming `Reader` API to access values in an Ion stream. This API is efficient, but has a rather steep learning curve and can often require verbose code. This PR creates a DOM-style API for loading values from an Ion stream into memory, allowing users to interact with them as they would any other JS value/Array/Object.

### `load()` and `loadAll()`

Two new top-level functions have been introduced:

```typescript
// Loads the first value from the provided data source (if present) and returns it
load(ionData: ReaderBuffer | Reader): Value | null

// Loads all of the values found in the provided data source and returns them
loadAll(ionData: ReaderBuffer | Reader): Value[]
```

### Javascript (`any`) interface

Each Ion data type is represented by a new class that extends an existing Javascript type:
* `dom.List` extends `Array`
* `dom.Integer` extends `Number`
* `dom.Timestamp` extends `Date`
* ....

This means that in most cases, you can treat values returned by `load()` the same way you would treat a standard Javascript equivalent.

```typescript
// Ion integer
let i: any = ion.load('7');
assert.equal(7, i);
assert.equal(8, i + 1);
assert.equal(49, i ** 2);
assert.equal(1, i / 7);
```

```typescript
// Ion string
let s: any = ion.load('"Saturn"');
assert.equal('Saturn', s);
assert.equal('Saturn' + ', Jupiter', s + ', Jupiter');
assert.equal('Saturn'.substr(4), s.substr(4));
assert.isNotNull(s.match(/atu/));
```

```typescript
// Ion list
let l: any = ion.load('[0, 2, 4, 6, 8, 10]');
assert.equal(4, l[2]);
assert.equal(6, l.length);
// iteration example
for (let value of l) {
    assert.isTrue(value % 2 == 0);
}
// Direct access to Array methods
assert.equal(30, l.reduce((sum, value) => sum + value, 0));
```

```typescript
// Ion struct
let person: any = ion.load(
  'person::{' +
      'name: {' +
          'first: "John", ' +
          'middle: "Jacob", ' +
          'last: "Jingleheimer-Schmidt",' +
     '},' +
     'age: 41' +
  '}'
);
assert.equal(41, person.age);
assert.equal(41, person['age']);
assert.equal('Jacob', person.name.middle);
for (let [key, value] of person) { // Iteration
    console.log(name + ': ' + value);
}
```

### Typescript (`Value`) interface

The Javascript API above is concise and convenient, but does not offer any static type safety. Each of the new classes in this PR also implements a new interface called `dom.Value`, which provides methods for indexing into Ion values as well as converting them to JS representations.

```typescript
// Ion integer
// `i` is an implementation of the Value interface. To use it, we'll need to
// be explicit about our coercions to JS types in order to satisfy 
// the Typescript compiler.
let i: Value = ion.load('7');

// The unary '+' operator will coerce our dom.Integer into a JS number
assert.equal(7, +i);
// We can explicitly convert this value to a JS number
assert.equal(7, i.numberValue()); 

// dom.Integer is not a text value, so there is no string value associated with it.
assert.isNull(i.stringValue());
```

```typescript
// Ion list
let l: Value = ion.load('[0, 2, 4, 6, 8, 10]');

// We can use the `get()` method to index into the Value
assert.equal(4, l.get(2));

// iteration example using the `values()` method
for (let value of l.values()!) {
    assert.isTrue(value % 2 == 0);
}
```

```typescript
// Ion struct
let person: Value = ion.load(
  'person::{' +
      'name: {' +
          'first: "John", ' +
          'middle: "Jacob", ' +
          'last: "Jingleheimer-Schmidt",' +
     '},' +
     'age: 41' +
  '}'
)!;
assert.equal(41, person.get('age'));
assert.equal('Jacob', person.get('name', 'middle'));

// Iteration example using the `fields()` method
for (let [key, value] of person.fields()) {
    console.log(name + ': ' + value);
}
```

### Recommended reading order

I suggest reading the following files first:
1. `src/dom/Value.ts`: All of the new classes introduced in this PR implement this interface.
2. `src/dom/DomValue.ts`: Serves as a common default implementation for the `Value` interface.
3. `src/dom/index.ts`: Defines the `load()` and `loadAll()` methods. 
4. `test/dom/dom.ts`: Demonstrates how to work with each data type after `load()`ing some Ion data.

The rest of the files contain class definitions for each Ion type and are pretty straightforward.

### Notes
1. I tested out a few different naming conventions before settling on using the plain Ion type names themselves grouped under the `dom` module. There are some name collisions that are easily remedied via package qualifiers or custom import aliases:
  * `ion.dom.String` vs `global.String`
  * `ion.dom.Boolean` vs `global.Boolean`
  * `ion.dom.Decimal` vs `ion.Decimal`
  * `ion.dom.Timestamp` vs `ion.Timestamp`
1. The provided implementation of `toString()` does not attempt to produce valid Ion. Overriding `toString()` in classes that extend `String` (i.e. `dom.String`, `dom.Symbol`) causes some `String` functionality (e.g. the `match` method) to break.

### Follow-on work
1. We still need to provide an `ion.dump(...)` method to perform the inverse of `ion.load()`. I held off on this to confirm that the API proposed in this PR is acceptable.
1. There are some `TODO`s in this code that will be turned into `ion-js` issues after the PR gets merged. Another PR will add the issue links to the comments.
1. There are some constructors that could/should take a wider variety of data types. For example, it should be possible to create a `dom.Decimal` from a `number`. However, this is not required for use of the `load()` API.
1. Once an `ion.dump()` method is available, we should design a formal integration with `ion-tests`.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

zslayton commented:
The tests are running fine, but it looks like there's an issue with `typedoc` generation. I'm looking into it.

almann commented on src/dom/Null.ts:
```diff
@@ -0,0 +1,50 @@
+ *      text encoding | Ion data type
+ *  ======================================
+ *      null          | IonType.NULL (implicit)
+ *      null.null     | IonType.NULL (explicit)
```
What is *implicit* vs. *explicit* mean here?  `null.null` and `null` are merely syntactic ways to spell the same thing, this doc seems to imply something more, but  I can't think of what could be meant.

almann commented on src/dom/Struct.ts:
```diff
@@ -0,0 +1,96 @@
+            })
+        }
+        this._setAnnotations(annotations);
+    }
```
I think this is burying the lede if I am reading this correctly.  You do surface field names as properties, yes?  If so, this should be documented explicitly in the doc comment.

What happens when they incidentally shadow methods or the internal fields?  Does this need to gate against that?

zslayton replied:
> You do surface field names as properties, yes?

Yes.

> If so, this should be documented explicitly in the doc comment.

Agreed, I'll add this.

> What happens when they incidentally shadow methods or the internal fields? Does this need to gate against that?

My plan (not implemented in this already quite large PR) is for all property accesses on structs to resolve to method names and internal fields first. If you have data that would shadow one of those fields, it can be accessed by `.get(fieldName)` instead. The iterators returned by `fields()` and `fieldNames()` would behave properly, returning the custom property values instead of the methods/fields.

```typescript
let s = ion.load('{stringValue: "Hello", foo: "bar"}')!; // The `stringValue` field shadows Value#stringValue()
let greeting = s.stringValue; // This resolves to the method handle and is probably not what the user wanted
let greeting = s['stringValue']; // This resolves to the method handle and is probably not what the user wanted
let greeting = s.stringValue(); // `greeting` is now "Hello"
let greeting = s.get('stringValue'); // `greeting is now "Hello"
```

This means that there are ~16 `Struct` field names that can have unexpected behavior, which seemed a better compromise than simply forbidding them.

almann commented on src/dom/Symbol.ts:
```diff
@@ -0,0 +1,27 @@
+
+    stringValue(): string {
+        return this.valueOf() as string;
+    }
```
Should this class override `toString` like `dom.String` does?

pbcornell commented on src/dom/Timestamp.ts:
```diff
@@ -0,0 +1,55 @@
+        );
+    }
+
+    timestampValue():ion.Timestamp {
```
add space before "ion.Timestamp"

zslayton replied:
Done.

pbcornell commented on test/dom/dom.ts:
```diff
@@ -0,0 +1,493 @@
+            ' 5e1' +
+            ' {a: 2.1, b: 4, c: [foo, (bar baz), "qux"]}' +
+            ' 1970-01-01T00:00:00.000Z'
+      )!;
```
is this `!` required?  `loadAll`'s signature doesn't specify `null` as a possible return;  maybe this is because we haven't enabled "strict" mode yet?

zslayton replied:
It's required in `strict` mode, yes. I originally wrote this as a separate lib with `strict` enabled and got used to it.

`loadAll` returns an array, so if there aren't any values it just returns an empty array. `load` returns the first value, so if there isn't one it returns `null`. Based on the `null` vs `throws` elsewhere, it should probably `throw` instead.

pbcornell commented on test/dom/dom.ts:
```diff
@@ -0,0 +1,493 @@
+
+   it('load() Null as Value', () => {
+      let n = ion.load('null.blob')!;
+
```
i'd suggest adding an assert based on `instanceof n`, too

zslayton replied:
Done.

pbcornell commented on test/dom/dom.ts:
```diff
@@ -0,0 +1,493 @@
+      assert.equal("Earth",s.get(2)!.stringValue());
+      assert.equal("Mars", s.get(3)!.stringValue());
+
+      let planets: string[] = [...s.values()!].map(s => s.stringValue()!);
```
simplify to: `[...s.values()!] --> s.values()!`  ?

zslayton replied:
`s.values()` returns an iterator and doesn't have a `map` method.

pbcornell commented on test/dom/dom.ts:
```diff
@@ -0,0 +1,493 @@
+         assert.isFalse(value.isNull());
+      }
+
+      assert.equal(2, [...s.fields()!].length)
```
`[...s.fields()!]` --> `s.fields()!` ?

zslayton replied:
`s.fields()` returns an iterator and so does not have a `length` property.

pbcornell commented on test/dom/dom.ts:
```diff
@@ -0,0 +1,493 @@
+         assert.isFalse(value.isNull());
+      }
+
+      assert.equal(2, [...s.fields()!].length)
```
`[...s.fields()!]` --> `s.fields()!` ?

zslayton replied:
Same comment RE: iterator.

zslayton commented on src/dom/DomValue.ts:
```diff
@@ -0,0 +1,127 @@
+            // This guarantees that users iterating over the fields of a struct or values in a list
+            // will see only Values from the source data or that they have created themselves.
+            Object.defineProperty(this, "_ionType", {enumerable: false});
+            Object.defineProperty(this, "_annotations", {enumerable: false});
```
Yeah, good call.

zslayton commented on src/dom/Value.ts:
```diff
@@ -0,0 +1,119 @@
+
+    /**
+     * Returns null if the Value is not an Integer, Float, or Decimal.
+     * Otherwise, returns the JS number representation of the Value.
```
👍 I like it.

zslayton commented on src/dom/Value.ts:
```diff
@@ -0,0 +1,119 @@
+     * Returns null if the Value is not a Timestamp.
+     * Otherwise, returns a Date representation of the Value.
+     */
+    dateValue(): Date | null;
```
The methods are named after their return types. `dateValue()` returns a native JS `Date` object. However, I'd meant to offer a `timestampValue(): Timestamp` along side it! I'll go back and add that.

zslayton replied:
> please check the first parameter of the dom-class constructors for consistency.  should others specify `protected readonly`? 

Adding `protected readonly` to a constructor parameter is a shorthand for creating a member field of the same name. We don't need to use it everywhere.

> looks like we're not using a "_" prefix consistently?

Good eye. I like the `protected readonly` shorthand for creating member fields, but disliked that this caused documentation and tooltips to include the `_` prefix so I seem to have mixed the two styles. I'll bite the bullet and use `_` everywhere.

> it's _entirely possible_ i don't understand the beauty of `[...x]` ;), but i expect `x` to be equivalent; am i missing out on the magic?

The `Value` interface's `fieldNames()`, `fields()`, and `values()` methods return `IterableIterator`s. While this is nice for iteration/looping, Javascript doesn't provide `Array`-style convenience methods like `map`, `filter`, and `reduce` for iterators. Likewise, there's no easy way to get their length.

The idiomatic JS solution to this problem is to use the spread operator to create an intermediate array you can perform these operations on. (`[...v.values()]`)

I had originally written `fieldNames()`, `fields()`, and `values()` to return arrays instead of iterators, but ran into a snag: `values()` is already a method on the `Array` type and it returns an `IterableIterator<any>`. I wasn't able to override it without also returning an iterator.

If you can suggest another name for `values()`, I can switch the return types to Arrays, probably making everyone's lives easier. `children()`?

pbcornell replied:
> I like the protected readonly shorthand for creating member fields, but disliked that this caused documentation and tooltips to include the _ prefix so I seem to have mixed the two styles. I'll bite the bullet and use _ everywhere.

If `_` is showing up in the documentation, I'd prefer we prevent that we some `@ignore` incantation, or fall back to not using the shorthand.  The documentation should correspond to the public API.

---

> If you can suggest another name for values(), I can switch the return types to Arrays, probably making everyone's lives easier. children()?

I think you're proceeding with `elements()`

zslayton commented on src/dom/Value.ts:
```diff
@@ -0,0 +1,119 @@
+     * Returns null if the Value is not a Blob or Clob.
+     * Otherwise, returns a Uint8Array representation of the Value's bytes.
+     */
+    uInt8ArrayValue(): Uint8Array | null;
```
I don't care for the name `byteValue()`, which suggests that it's going to return a single byte. I used `uInt8ArrayValue()` for consistency with the other methods in this interface, which follow the (boring, explicit) `returnType`+`Value` format.

pbcornell replied:
Agreed.  I'm in favor of boring, explicit, and _clear_.  Makes me wonder whether we should deprecate `Reader.byteValue()` in favor of a new `Reader.uInt8ArrayValue()` method...?

zslayton commented on src/dom/Value.ts:
```diff
@@ -0,0 +1,119 @@
+     * Otherwise, will attempt to index into the Value with each pathElement in turn
+     * until:
+     *  - A value is found that cannot use the current PathElement as an index.
+     *    (e.g. The current Value is a List but the current PathElement is a number.)
```
> do you mean "current PathElement is not a number"? 
D'oh, yep.

zslayton commented on src/dom/Value.ts:
```diff
@@ -0,0 +1,119 @@
+     * Returns null if the Value is not an instance of the class type represented by the provided
+     * constructor. Otherwise, casts the Value as type 'T' and returns it.
+     */
+    as<T extends Value>(ionValueType: Constructor<T>): T | null;
```
For users leveraging type information, casting a `Value` to its concrete implementation (e.g. `Struct`) can enable much more ergonomic interactions.

```typescript
let v: Value = ion.load('{foo: bar, baz: 42}');
for (let value of v.values()!) {
  // ...
}
numEntries = [...v.fields()].length;
sum = [...v.fields()].reduce((sum, value) => sum + value, 0);

// Turn our `Value` into a `Struct`
let l = v.as(Struct)!;

// Now the compiler knows it is-an Array
for (let value of l) {
  // ...
}
numEntries = l.length;
sum = l..reduce((sum, value) => sum + value, 0);
```

zslayton commented on src/dom/index.ts:
```diff
@@ -0,0 +1,154 @@
+import {BinaryReader} from "../IonBinaryReader";
+
+/**
+ * Reads the provided ion data source into memory, constructing Value objects to represent
```
Fixed.

zslayton commented on src/dom/index.ts:
```diff
@@ -0,0 +1,154 @@
+}
+
+/**
+ * Reads the first value from the provided ion data source into memory, constructing
```
Fixed.

zslayton commented on src/dom/index.ts:
```diff
@@ -0,0 +1,154 @@
+
+/**
+ * Reads the first value from the provided ion data source into memory, constructing
+ * an Value object to represent it.
```
At some point I did a sweeping rename of `IonValue` -> `Value`, `IonStruct` -> `Struct`, etc. The fallout was a number of these dangling `a`/`an` conflicts. :( Thanks for pointing them out!

Fixed.

zslayton commented on src/dom/index.ts:
```diff
@@ -0,0 +1,154 @@
+    return ionType === null ? null : loadValue(reader);
+}
+
+function createReader(ionData: ReaderBuffer | Reader): Reader {
```
Done.

zslayton commented on src/dom/index.ts:
```diff
@@ -0,0 +1,154 @@
+        case IonTypes.INT:
+            return reader.intSize() == IntSize.Number
+                ? new Integer(reader.numberValue()!, annotations)
+                :  new Integer(reader.bigIntValue()!, annotations);
```
Fixed.

zslayton commented on src/dom/Blob.ts:
```diff
@@ -0,0 +1,24 @@
+    /**
+     * Constructor.
+     * @param data          Raw, unsigned bytes to represent as a blob.
+     * @param annotations   An optional array of strings to associate with `bytes`.
```
`data` was originally named `bytes` -- a refactor missed the doc comment. Fixed.

zslayton commented on src/dom/Boolean.ts:
```diff
@@ -0,0 +1,44 @@
+ *          // this code WILL be executed
+ *      }
+ *
+ *
```
Done.

zslayton commented on src/dom/Decimal.ts:
```diff
@@ -0,0 +1,43 @@
+     * @param annotations   An optional array of strings to associate with `value`.
+     */
+    constructor(value: ion.Decimal, annotations: string[] = []) {
+        super(...[value.getCoefficient(), value.getExponent(), value.isNegative()]);
```
It cannot. The parent constructor (a mixin) is required to accept a single spread parameter. Attempting to pass it 3 distinct values results in the following error:

> TS2554: Expected 0-1 arguments, but got 3.

zslayton commented on src/dom/Decimal.ts:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    // Required to produce sensible output for toString().
+    // Without it, toString() relies on the underlying Decimal value's coefficient.
```
Good catch. I rewrote `toString()` to address this and forgot to remove the comment. Removed.

zslayton commented on src/dom/Integer.ts:
```diff
@@ -0,0 +1,56 @@
+     */
+    constructor(value: JSBI | number, annotations: string[] = []) {
+        // If the provided value is a JS number, we will defer constructing a BigInt representation
+        // of it until it requested later by a call to bigIntValue().
```
Removed the 'it'.

zslayton commented on src/dom/Integer.ts:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    bigIntValue(): JSBI {
+        if(this._bigIntValue === null) {
```
Done.

zslayton commented on src/dom/Null.ts:
```diff
@@ -0,0 +1,50 @@
+ *
+ * An Ion null differs from a Javascript null in that an instance of an Ion null may have data associated with it.
+ *
+ * In particular, Ion nulls have an Ion data type:
```
In almost all cases, users will not check to see if the underlying implementation of `Value` is a `dom.Integer` or a `dom.Null`.

Requiring each class to represent Ion's `null` means branching on the `null` possibility for all functions in all concrete classes. Having a separate `dom.Null` class simplifies the implementation appreciably while still allowing for predictable behavior from `isNull()`, `getType()`, and the various `*Value()` methods on each class.

zslayton commented on src/dom/Sequence.ts:
```diff
@@ -0,0 +1,51 @@
+            let children = this as Value[];
+            let maybeChild: Value | undefined = children[pathHead];
+            let child: Value | null = maybeChild === undefined ? null : maybeChild;
+            if (pathTail.length === 0 || child === null) {
```
In that case, `child`'s type is `Value | undefined` while the method's return type is `Value | null`.

zslayton commented on src/dom/Sequence.ts:
```diff
@@ -0,0 +1,51 @@
+        }
+
+        toString(): string {
+            return '[' + this.join(', ') + ']';
```
That's an easy enough change. However, `toString()` doesn't attempt to produce valid Ion. (Note that it doesn't include annotations, for example.)

zslayton commented on src/dom/Struct.ts:
```diff
@@ -0,0 +1,96 @@
+     */
+    constructor(fields: IterableIterator<[string, Value]>, annotations: string[] = []) {
+        super();
+        for(let [fieldName, value] of fields) {
```
Done.

zslayton commented on src/dom/String.ts:
```diff
@@ -0,0 +1,27 @@
+export class String extends DomValue(global.String, IonTypes.STRING) {
+    /**
+     * Constructor.
+     * @param _text          The text value to represent as a string.
```
Done.

zslayton commented on src/dom/Struct.ts:
```diff
@@ -0,0 +1,96 @@
+                enumerable: true,
+                value: value,
+                writable: true
+            })
```
Done.

zslayton commented on src/dom/Struct.ts:
```diff
@@ -0,0 +1,96 @@
+
+    toString(): string {
+        return '{'
+            + [...this.fields()]
```
`this.fields()` returns an iterator, which does not have a `map` or `join` method. Using `[...this.fields()]` creates an intermediate array and (though clunky and less efficient) is considered idiomatic.

zslayton commented on src/dom/Symbol.ts:
```diff
@@ -0,0 +1,27 @@
+    /**
+     * Constructor.
+     * @param symbolText    The text to represent as a symbol.
+     * @param annotations   An optional array of strings to associate with this null value.
```
Yep, fixed.

zslayton commented on src/dom/Timestamp.ts:
```diff
@@ -0,0 +1,55 @@
+    /**
+     * Constructor.
+     * @param dateOrTimestamp   A `Date` or `Timestamp` to represent as a timestamp.
+     * @param annotations       An optional array of strings to associate with this null value.
```
Fixed.

zslayton commented on src/dom/Timestamp.ts:
```diff
@@ -0,0 +1,55 @@
+        } else {
+            timestamp = dateOrTimestamp;
+            date = timestamp.getDate();
+
```
Done.

zslayton commented on src/dom/Value.ts:
```diff
@@ -0,0 +1,119 @@
+     * @param pathElements  One or more values to be used to index into the Value.
+     * @returns null if no value is found at the specified path. Otherwise, returns the discovered Value.
+     */
+    get(...pathElements: PathElement[]): Value | null;
```
> If pathElements resolves to a null value, I'm expecting an instance of Value to be returned for which isNull() returns true.

Can you elaborate on this? If there isn't an Ion value at the specified path, why would an instance of `Value` be returned?

zslayton replied:
To clarify the current behavior:

> If pathElements resolves to a null value

* If `pathElements` resolves to an Ion null, `get` will currently return a `Value` for which `isNull()` is true and `getType()` returns the correct Ion type.
* If `pathElements` resolves to a JS `null`, it means that no value was found at the specified path.

zslayton commented on src/dom/index.ts:
```diff
@@ -0,0 +1,154 @@
+ *
+ * If `ionData` is a ReaderBuffer, a new Reader will be created to process it.
+ * If it is already a Reader, next() will be called once at the top level to consume the
+ * first value found in the data.
```
> Wondering whether this should align with the writeValue()/writeValues() handling of a passed Reader instance. Iirc, a reader's next() is not called if the reader is positioned on a value.

That's an easy enough change. It's worth being consistent when possible.

> "will be called once at the top level": is there an assumption that a passed reader is at the top level? what if it isn't?

There's no requirement that it be at the top level, I just flubbed the phrasing. I originally wrote "will be called once" and then thought "...but next() can be called a lot more than once if the data has nested containers" and tried to qualify it. I'll rephrase.

zslayton commented on src/dom/index.ts:
```diff
@@ -0,0 +1,154 @@
+    let children: Value[] = [];
+    reader.stepIn();
+    let ionType: IonType | null = reader.next();
+    while (ionType !== null) {
```
Done.

zslayton commented on src/dom/index.ts:
```diff
@@ -0,0 +1,154 @@
+    let annotations: string[] = reader.annotations();
+    reader.stepIn();
+    let ionType: IonType | null = reader.next();
+    while (ionType !== null) {
```
Done.

zslayton commented on src/dom/index.ts:
```diff
@@ -0,0 +1,154 @@
+        return ionData as Reader;
+    }
+    return makeReader(ionData as ReaderBuffer);
+}
```
I remember the original challenge I encountered here. `ReaderBuffer` is comprised of:
* `string`
* `ReaderOctetBuffer`
  * `ArrayBufferLike`
    * `ArrayBuffer`
      * `ArrayBufferView`
      * ... (different view sizes)
    * `SharedArrayBuffer`
    * ...
  * `ArrayLike<number>`

Many of these are interfaces or type aliases and do not exist at runtime. There's no comprehensive list of concrete types we can test for with (e.g.) `typeof` or `instanceof`.

I'm going to leave this as-is for now. I suggest providing a public `AbstractReader` class as a common base class for `Reader` implementations, even if it offers no concrete functionality. Then we can simply test `instanceof AbstractReader` to make this distinction.

zslayton commented on src/dom/index.ts:
```diff
@@ -0,0 +1,154 @@
+        ionType = reader.next();
+    }
+    reader.stepOut();
+    return new Struct(children.entries(), annotations);
```
Regarding duplicate field names:

We could modify the `Struct` class to have both an `Array` of the fields passed to the constructor and a `Map` containing the most recently updated value for each key.
* Property accesses via `.foo`, `['foo']` or `.get('foo')` would return the last value.
* The `fields()` and `values()` iterators could accept a flag indicating that ALL fields should be returned, not just the last one for each name.
* If there's demand for it, later on we could provide a `getAll(fieldName)`. I don't expect this to be common, however.

However, this raises the question of what to do when a user sets a field. Do we append that field to the end of the array, acting as though it's tacked on to the end of the struct? Users may be surprised if they serialize it and find duplicates in the output.

zslayton commented:
### Updates

The [most recent commit](https://github.com/amzn/ion-js/pull/557/commits/531cbd89b6fb584093931f2317b42c859e383d83) includes the following changes:

* Renamed `Value`'s `_annotations` field to `_ionAnnotations`.
* `Value`'s `[type]Value()` functions will throw an `Error` if the requested operation is not supported by the Value's ionType.
* `Null` now contains type checking logic to see if a conversion should succeed but return null (e.g. `ion.load('null.string').stringValue() === null`) or if it should fail due to a type mismatch (e.g. `ion.load('null.int').stringValue()` produces an `Error`)
* Added a `timestampValue(): Timestamp | null` function alongside `dateValue()` for lessless access to ion Timestamp values.
* Renamed the `values()` method to `elements()` so it no longer shadows [`Array#values`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/values), which returns an iterator.
* `Value`'s `fields()`, `fieldNames()`, and `elements()` methods return arrays instead of iterators. This allows users to call `map`, `filter`, `join`, etc on the results of those methods natively instead of collecting their output into an intermediate array via `[...value.elements()]` first. Somewhat surprisingly, JS values often return an array for methods you might expect to return an iterator. For examples, see:
  *  [`Object#entries`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries)
  * [`Object#keys`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/keys)
  * [`Object#values`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/values)
* `ion.load()` will only call next on the provided Reader if it is not already positioned over a value.
* Several classes no longer store their constructor parameters as a member field. In some cases it may have offered a micro-optimization when converting to a JS primitive, but it required an amount of complexity that we can hold off on until we need it.
* Fixed a number of spacing issues/typos that Peter identified.

### Blocking issue
I've identified [a bug in `typedoc`'s mixin support](TypeStrong/typedoc#1207) that is currently causing the build to fail. We should discuss short term workarounds to avoid unnecessary delays.

zslayton commented:
I've unblocked the build by creating [a temporary `typedoc` fork](https://github.com/zslayton/typedoc/) with [a patch](https://github.com/zslayton/typedoc/commit/7f58d79b0a445434e487d6b61bb14933fbe0c329). `package.json` has been tweaked to point to the fork until a fix is available.

zslayton commented:
The `typedoc` patch was merged and [released in v0.16.10](https://github.com/TypeStrong/typedoc/pull/1208) over the weekend. (Thanks, @Gerrit0!)

I've updated the `typedoc` dependency to point to the new version and deleted my temporary fork.

pbcornell commented on src/dom/Value.ts:
```diff
@@ -0,0 +1,126 @@
+    stringValue(): string | null;
+
+    /**
+     * For the Timestamp type, returns a JS Date representation of the Value; otherwise throws
```
Let's indicate that this may result in data loss

zslayton replied:
Good call, added a note.

zslayton commented on src/dom/index.ts:
```diff
@@ -0,0 +1,159 @@
+    let reader = _createReader(ionData);
+    let ionValues: Value[] = [];
+    let ionType: IonType | null = reader.next();
+    while (ionType !== null) {
```
Much more concise, thanks.

pbcornell commented on src/dom/Null.ts:
```diff
@@ -41,10 +61,72 @@ export class Null extends DomValue(Object, IonTypes.NULL) {
+
+    // If this Null's Ion type supports the requested operation, throw an Error indicating this was a null dereference.
+    // Otherwise, throw an Error indicating that the requested operation is not supported.
+    private _unsupportedOperationOrNullDerefence(operation: string): never {
```
Derefence --> Dereference

zslayton replied:
🤦‍♂ Fixed!

pbcornell commented on src/dom/Null.ts:
```diff
@@ -41,10 +61,72 @@ export class Null extends DomValue(Object, IonTypes.NULL) {
+        if (Null._operationIsSupported(this.getType(), operation)) {
+            return null;
+        }
+        throw new Error(`'${operation}' is not supported by Ion type ${this.getType().name}`);
```
drop single quotes, add `()`?  as in:  `${operation}() is not supported...`

zslayton replied:
Sure.

pbcornell commented on src/dom/Null.ts:
```diff
@@ -41,10 +61,72 @@ export class Null extends DomValue(Object, IonTypes.NULL) {
+    // Otherwise, throw an Error indicating that the requested operation is not supported.
+    private _unsupportedOperationOrNullDerefence(operation: string): never {
+        if (Null._operationIsSupported(this.getType(), operation)) {
+            throw new Error(`${operation} called on a null ${this.getType().name}.`);
```
add `()`, as in:  `${operation}() called...`

zslayton replied:
Done.

pbcornell commented on src/dom/Null.ts:
```diff
@@ -41,10 +61,72 @@ export class Null extends DomValue(Object, IonTypes.NULL) {
+        if (Null._operationIsSupported(this.getType(), operation)) {
+            throw new Error(`${operation} called on a null ${this.getType().name}.`);
+        }
+        throw new Error(`'${operation}' is not supported by Ion type ${this.getType().name}`);
```
drop single quotes around `${operation}` ?

zslayton replied:
Done.

zslayton commented on src/dom/Struct.ts:
```diff
@@ -5,6 +5,11 @@ import {DomValue} from "./DomValue";
+ *    let s: any = ion.load('{foo: 1, bar: 2, baz: qux::3}');
+ *    assert.equal(6, s.foo + s['bar'] + s.baz);
+ *
  * [1] http://amzn.github.io/ion-docs/docs/spec.html#struct
```
I have a PR brewing that will introduce the new methods-take-precedence behavior. I'll be sure to spell out that behavior in the updated docs.

zslayton commented on src/dom/Null.ts:
```diff
@@ -26,9 +28,27 @@ import {IonType, IonTypes} from "../Ion";
+        ['fields', new Set([IonTypes.STRUCT])],
+        ['fieldNames', new Set([IonTypes.STRUCT])],
+        ['elements', new Set([IonTypes.LIST, IonTypes.SEXP, IonTypes.STRUCT])]
+    ]);
```
I'm going to see if I can add the supported types as a property directly on the methods' `Function` objects. I'll do so in an upcoming PR.
